### ISSUE TITLE
[Feature]: Keymapping Tables

### ISSUE THREAD

**Issue by LandonSchropp** [LunarVim/LunarVim#1548]
### Feature motivation

This is a follow up to #1525. The problem I was experiencing in that issue was that some of my keyboard mappings were randomly not applied 50% of the time. Here are a few examples of settings that were only working half the time:

``` lua
lvim.keys.normal_mode["H"] = "gT"
lvim.keys.normal_mode["L"] = "gt"
lvim.builtin.telescope.defaults.mappings.i["<cr>"] = actions.file_tab
```

After a ton of digging, I finally figured out what was going on. Internally, LunarVim was using different forms of these mappings:

[keymappings.lua](https://github.com/LunarVim/LunarVim/blob/rolling/lua/keymappings.lua#L105-L106):

 ``` lua
["<S-l>"] = ":BufferNext<CR>",
["<S-h>"] = ":BufferPrevious<CR>",
```

[telescope.lua](https://github.com/LunarVim/LunarVim/blob/rolling/lua/core/telescope.lua#L58):

``` lua
["<CR>"] = actions.select_default + actions.center
```

When my config is loaded by LunarVim and converted to NeoVim mappings, the iteration order seems to be non-deterministic, which causes my mappings to be randomly applied.

### Describe the expected solution

I'd propose a special subclass of Lua's table that can handle keymappings. I'd expect this class to pick a standard for keys (both in terms of casing and modifiers), and to be used anywhere keymappings are used. The only two places that I've encountered are in the normal `lvim.keys.<mode>` maps, as well as `telescope.defaults.mappings`, but there could be others.

Since mappings are not case (or form) sensitive in Neovim, this would follow the editor's normal behavior.

I would expect this to be able to handle the following cases (and maybe more I'm not considering):

``` lua
-- These should all be assigned to the *same value* inside the table. They should override each other.
lvim.keys.normal_mode["B"] = ":echo Banana<cr>"
lvim.keys.normal_mode["<S-b>"] = ":echo Basketball<cr>"
lvim.keys.normal_mode["<s-b>"] = ":echo Binoculars<cr>"

-- Each one of these should return the exact same value.
lvim.keys.normal_mode["B"] -- ":echo Binoculars<cr>"
lvim.keys.normal_mode["<S-b>"] -- ":echo Binoculars<cr>"
lvim.keys.normal_mode["<s-b>"] -- ":echo Binoculars<cr>"

-- When iterating over the keys, the mapping should only be selected once. I don't think it matters
-- which format is used for key by LunarVim, as long as it's programmatically consistent. For
-- example, the following should output something like this:
--
-- B: :echo Binoculars<cr>
for key, value in pairs(lvim.keys.normal_mode) do
  print(key .. ": " .. value)
end
```

### Describe the alternatives you've considered

I could update my config to match the internal implementations (which is probably how I'll resolve my immediate issue). However, I don't think anyone should have to know how the keymappings are defined in order to override them. Furthermore, the keymappings are not defined consistently internally (and they shouldn't need to be.) This will likely be a confusing mess for anyone else who tries to override LunarVim case-sensitive or modifier mappings and happens to pick the wrong form.

### Additional context

This took me several days of debugging to identify the root cause. 😔 Had I not found the root cause, I would have had to drop LunarVim as my editor to maintain my productivity.

**Comment by younger-1** [comment]
A bit of insight of lunarvim from my point of view:
```md
- require("bootstrap"):init()
  - update runtime-dir
  - require("impatient").setup()
  - require("config"):init { config_path }
    - require("config.defaults")
      - lvim
      - lvim.lang
    - require("core.builtins").config()
      - require("keymappings").config()
        - lvim.keys                        <------------------------------- Builtin keymappings defined, but not sent to nvim for now
      - require("core.which-key").config()
      - require("core.gitsigns").config()
      - require("core.cmp").config()
      - require("core.dashboard").config()
      - require("core.dap").config()
      - ...
    - require("config.settings").load_options()
  - require("plugin-loader"):init { path }

- require("config"):load()
  - local autocmds = require("core.autocmds")
    - lvim.autocommands
  - luafile config_path                <------------------------------- Here, your personal lvim.keys.normal_mode["<S-b>"] be defined
  - autocmds.define_augroups(lvim.autocommands)
  - require("config.settings").load_commands()
- local plugins = require "plugins"
  - { config = require("core.[]").setup() }
- require("plugin-loader"):load { plugins, lvim.plugins } 
                      <------------------------- bufferline plugin overrides your personal lvim.keys.normal_mode["<S-b>"]
                                                              See: https://github.com/LunarVim/LunarVim/blob/rolling/lua/core/bufferline.lua#L9-L10
- require("core.log")
- vim.cmd("colorscheme " .. lvim.colorscheme) -- Colorscheme must get called after plugins are loaded or it will break new installs.
- require("utils").toggle_autoformat()
- require("core.commands").load(commands.defaults)
- require("lsp").global_setup()
- require("keymappings").setup()          <------------------------------- All keymappings are sent to nvim throuth nvim.api
```

**Comment by LandonSchropp** [comment]
@younger-1 Thanks for the reply, but I'm a bit confused. The issue here is that the casing on the keybindings can use multiple definitions of the binding in the same table in the `lvim` config. For example, it's possible to have both `H` and `<S-h>` in the same `normal_mode` table, because the keys are unique. This causes *both* keymaps to get loaded by `keymappings.lua`, in a non-deterministic order. And whichever one happens to get loaded last wins.

**Comment by younger-1** [comment]
@LandonSchropp Oh, I got it. Had to say it is a bit confusing.

**Comment by LandonSchropp** [comment]
Sorry, my original post probably wasn't clear. 🙂

I'm happy to put out a draft PR with an example of what I was thinking if that'd help.

**Comment by younger-1** [comment]
Yep! It is a good thing to reveal the strange little inner stuff for us.

**Comment by kylo252** [comment]
It's hard to balance-out the ability to provide some sane-defaults while still have them overridable. I personally did not know that you can use `H`. Furthermore, this is one of the main reasons why I made `<leader>Lk` to show all the keybindings that aren't managed by which-key <leader> prefix.

I understand the frustration, but there some subtle things that make this more complicated than what it appears to be on the surface imo. 

I have been toying around with the idea of using https://github.com/b0o/mapx.nvim, do you guys think it would solve a problem, or is just another thing to learn?

**Comment by b0o** [comment]
> I have been toying around with the idea of using https://github.com/b0o/mapx.nvim, do you guys think it would solve a problem, or is just another thing to learn?

I strongly encourage using Mapx! :wink: :wink:

If there are any features or changes that'd help you out, drop me an issue. I'd be happy assist.

**Comment by LandonSchropp** [comment]
MapX would definitely solve my issue (assuming it would replace the existing configuration).

**Comment by gonzaloserrano** [comment]
Without knowing any LunarVim internals, would an API to disable default mappings help here? I'm being hit here at its a PITA (opened and then closed #1960 as a dupe of this one).

**Comment by gonzaloserrano** [comment]
Also, I'm not sure about the PR title: this is not a feature, but a bugfix.
